Realm/RLMObjectBase.mm

```diff
@@ -159,32 +156,67 @@ - (NSArray *)linkingObjectsOfClass:(NSString *)className forProperty:(NSString *
     }
     
     size_t col = prop.column;
-    NSUInteger count = _row.get_backlink_count(*table, col);
+    NSUInteger count = object->_row.get_backlink_count(*table, col);
     NSMutableArray *links = [NSMutableArray arrayWithCapacity:count];
     for (NSUInteger i = 0; i < count; i++) {
-        [links addObject:RLMCreateObjectAccessor(_realm, schema, _row.get_backlink(*table, col, i))];
+        [links addObject:RLMCreateObjectAccessor(object->_realm, schema, object->_row.get_backlink(*table, col, i))];
     }
     return [links copy];
 }
 
-- (id)objectForKeyedSubscript:(NSString *)key {
-    if (_realm) {
-        return RLMDynamicGet(self, key);
+id RLMObjectBaseObjectForKeyedSubscript(RLMObjectBase *object, NSString *key) {
+    if (!object) {
+        return nil;
+    }
+
+    if (object->_realm) {
+        return RLMDynamicGet(object, key);
     }
     else {
-        return [self valueForKey:key];
+        return [object valueForKey:key];
     }
 }
 
-- (void)setObject:(id)obj forKeyedSubscript:(NSString *)key {
-    if (_realm) {
-        RLMDynamicValidatedSet(self, key, obj);
+void RLMObjectBaseSetObjectForKeyedSubscript(RLMObjectBase *object, NSString *key, id obj) {
+    if (!object) {
+        return;
+    }
+
+    if (object->_realm) {
+        RLMDynamicValidatedSet(object, key, obj);
     }
     else {
-        [self setValue:obj forKey:key];
+        [object setValue:obj forKey:key];
     }
 }
 
+
+BOOL RLMObjectBaseAreEqual(RLMObjectBase *o1, RLMObjectBase *o2) {
+    // if not the correct types throw
+    if ((o1 && ![o1 isKindOfClass:RLMObjectBase.class]) || (o2 && ![o2 isKindOfClass:RLMObjectBase.class])) {
+        @throw RLMException(@"Can only compare objects of class RLMObjectBase");
+    }
+    // if identical object
+    if (o1 == o2) {
+        return YES;
+    }
+    // if one is nil
+    if (!!o1 != !!o2) {
```

Comment:
Also note they are not exactly the same - we are returning NO only if one is nil and the other is not. Your suggested code only works because we previously checked if the two objects are equal, which is true if they are both nil. In the case that these lines were reordered your code would be be incorrect.